### ISSUE TITLE
Add no-unknown-custom-media

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#6362]
### What is the problem you're trying to solve?

I'd like to disallow unknown custom media queries. For example:

```css
@custom-media --foo (max-width: 30em);
@media (--bar) {}
```

### What solution would you like to see?

A new rule to catch this. Like https://github.com/stylelint/stylelint/issues/6361, it's currently [available as a plugin](https://github.com/csstools/stylelint-media-use-custom-media) but it [(perhaps) meets our criteria](https://stylelint.io/user-guide/rules/about) to be a built-in rule. Unlike https://github.com/stylelint/stylelint/issues/6361, it's for [a draft spec](https://drafts.csswg.org/mediaqueries-5/#custom-mq). However, we already have a custom media queries rule ([custom-media-pattern](https://stylelint.io/user-guide/rules/list/custom-media-pattern)) and both [PostCSS Preset Env](https://preset-env.cssdb.org/features/#custom-media-queries) and [Lightening CSS](https://github.com/parcel-bundler/lightningcss#features) offer syntax lowering for it.

Interestingly, the plugin also enforces using custom media queries through its [`always` and `always-known` primary options](https://github.com/csstools/stylelint-media-use-custom-media#options). Shall we take the same approach or create two rules?:

- one that checks for unknown custom media queries 
- and another that ensures matching queries use the corresponding name 

I'm leaning towards the latter.

- Name: `no-unknown-custom-media`
- Primary option: `true`
- Secondary options: `importForm` & `resolver`?
- Autofixable: No
- Message: - `Unexpected unknown custom media`
- Description: "Disallow unknown custom media queries."
- Extended description: "This rule considers custom media queries defined within the same source to be known. Use the `importForm` secondary option to specify more sources."
- Section: "Avoid errors" -> "General"

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the proposal. The new rule sounds good to me. 👍🏼 

> Shall we take the same approach or create two rules?:

I agree with creating two rules; one is `no-unknown-custom-media` (avoiding errors), while another would be something like `custom-media-***` (enforcing conventions). Although, it seems unclear whether the latter rule would be needed now.

**Comment by jeddy3** [comment]
> I agree with creating two rules; one is `no-unknown-custom-media` (avoiding errors), while another would be something like `custom-media-***` (enforcing conventions). Although, it seems unclear whether the latter rule would be needed now.

SGTM.

- Name: `no-unknown-custom-media`
- Primary option: `true`
- Secondary options: `importFrom: []` - array of glob paths to `.css` files
- Autofixable: No
- Message: - `Unexpected unknown custom media`
- Description: "Disallow unknown custom media queries."
- Extended description: "This rule considers custom media queries defined within the same source to be known. You can use the `importFrom` secondary option to specify more sources."
- Section: "Avoid errors" -> "General"

Labelling as ready to implement.

If anyone fancies [contributing](https://stylelint.io/contributing) this rule, there are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

**Comment by ybiquitous** [comment]
@fpetrakov Thank you for starting this issue! @romainmenke's package seems to provide `parseCustomMedia()`, so I believe there are no blockers. 👍🏼

**Comment by ybiquitous** [comment]
FYI, we cannot provide the `importFrom` secondary option because of a resolver problem. See also:

- #4088 
- #6746

**Comment by silverwind** [comment]
`@custom-media` seems to not be part of any CSS standard, I wonder why stylelint chose to even support it.

**Comment by romainmenke** [comment]
> @custom-media seems to not be part of any CSS standard

It is part of this standard : https://www.w3.org/TR/mediaqueries-5/#custom-mq
Browsers are however not very eager to implement this (yet).

**Comment by silverwind** [comment]
Yeah, I guess they are more likely to just make variables work inside media queries than to implement another at-rule, I suppose :)

**Comment by romainmenke** [comment]
No, they are even less eager to do that :D

But I should take the time in the near future to check if it can get unstuck. It is also the 4th most visited page on the `postcss-preset-env` repo.

<img width="490" alt="Screenshot 2024-04-23 at 19 45 48" src="https://github.com/stylelint/stylelint/assets/11521496/2bf27227-768f-449c-9ae4-06c119d757a7">
